### PR TITLE
Api exercise

### PR DESCRIPTION
**Pull request by ralucanegoita10**

### CODE REVIEW

File: source/data/game.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[
```
Comment (edwardbrowncross) [inline source/data/game.json:1]: We'd normally use our github repository for only storing the code of our project. The data it generates while running (like logs, or scores or user data) would be stored a different way if it needs to be stored at all.

File: source/frontend/connect4.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+$.ajax({
+  type: 'GET',
+  url: '/get-board',
+  contentType: 'application/json',
+  success: (result) => {
+    drawBoard(result.board);
+
+    $('h3').text(`${player1}: it is your turn, please pick a column to drop your chip.`).show();
+
+    const playerObj1 = result.scores.filter((item) => item.player === 1)[0];
```
Comment (edwardbrowncross) [inline source/frontend/connect4.js]: You could also use `.find` for this
```suggestion
    const playerObj1 = result.scores.find((item) => item.player === 1);
```

File: source/frontend/connect4.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      } else if (board[row][column] === 1) {
+        $(`#row-${row}-column-${column}`).css('background-color', 'red');
+      } else if (board[row][column] === 2) {
+        $(`#row-${row}-column-${column}`).css('background-color', 'yellow');
+      }
+    }
+  }
+}
+function setNextTurn(player) {
+  if (player === 1) {
```
Comment (edwardbrowncross) [inline source/frontend/connect4.js]: Another way to write this and prevent some duplication would be:
```js
const playerName = player === 1 ? player1 : player2;
$('h3').text(`${playerName}: it is your turn, please pick a column to drop your chip.`).show();
```

File: source/frontend/connect4.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    $('h3').text(`${player1}: it is your turn, please pick a column to drop your chip.`).show();
+  } else {
+    $('h3').text(`${player2}: it is your turn, please pick a column to drop your chip.`).show();
+  }
+}
+
+function gameEnd(winningPlayer) {
```
Comment (edwardbrowncross) [inline source/frontend/connect4.js:65]: There's a little bit of inconsistency between this function and the previous one. `setNextTurn` takes a number (1 or 2). But `gameEnd` takes the name of the player. Looking at where they're used, I see how that makes the functions simpler right now, but consistency in how information is passed into function makes code easier to understand and use in the future

File: source/frontend/connect4.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  $('h3').hide();
+  $('h2').hide();
+  $('h1').text(`${winningPlayer} has won! Press "New game" to start again!`)
+    .css('fontSize', '30px')
+    .show();
+  gameEnded = true;
+}
+
+function tryThis() {
```
Comment (edwardbrowncross) [inline source/frontend/connect4.js]: This function takes up half the file and yet its name doesn't make it very clear what it does. Names like `setupEvents` or `initEventListeners` or `addClickHandlers` or similar would be more descriptive

File: source/frontend/functions.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+const turnCount = 0;
+let currentPlayer;
+
+const Game = {
+
+  // firstPlayer: 'yellow',
```
Comment (edwardbrowncross) [inline source/frontend/functions.js]: If this code isn't needed any more, then it can be removed. If you need it again in the future, you can always go and find it in your commit history.

File: source/frontend/functions.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  // winMsg: 'The winner is: ',
+  // countToWin: 4,
+  boardLength: 7,
+  boardHeight: 6,
+};
+
+let player1;
+let player2;
+
+function getBoard(_board) {
```
Comment (edwardbrowncross) [inline source/frontend/functions.js]: I can't see anywhere this function is used. It looks to duplicate some of the functionality now in `drawBoard`. It sort of feels as though this file ought to be merged with `connect4.js` as I don't see a lot of value in this being separate

File: source/frontend/index.html [hunk 8]
```diff
@@ -22,8 +22,20 @@ <h1 class="title">Welcome to Connect4</h1>
                 <h2>Connect 4 chips to win !!</h2>
                
                 <h3>Let's start</h3>
+            
+            </div>
 
+            <br>
+            
+            <div id="Red">
```
Comment (edwardbrowncross) [inline source/frontend/index.html:30]: There's some inconsistent indentation here.

File: source/frontend/index.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+$(() => {
```
Comment (edwardbrowncross) [inline source/frontend/index.js]: No longer needed?

File: source/server.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+/* Create express server instance */
+const app = express();
+
+let scores;
+
+app.use(express.static('./source/frontend'));
+app.use(express.json());
+
+const fs = require('fs').promises;
```
Comment (edwardbrowncross) [inline source/server.js]: It's standard practice to put all the imports (requires) at the top of the file. It makes it really clear up-front what dependencies the file has.

File: source/server.js [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+  [0, 0, 0, 0, 0, 0, 0],
+  [0, 0, 0, 0, 0, 0, 0],
+  [0, 0, 0, 0, 0, 0, 0],
+  [0, 0, 0, 0, 0, 0, 0],
+];
+
+/* Reads from file and adds contents to global variable 'scores' */
+async function readScores() {
+  const rawData = await fs.readFile('./source/data/game.json', 'utf-8');
+  const parsedElement = JSON.parse(rawData);
```
Comment (edwardbrowncross) [inline source/server.js]: These two lines could be condensed to one
```suggestion
scores = JSON.parse(rawData);
```

File: source/server.js [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+  scores = parsedElement;
+}
+
+/* Selects the element of the file that matches the given player
+   and updates its score */
+async function updateScore(player) {
+  try {
+    await readScores();
+
+    const playerObj = scores.filter((item) => item.player === player)[0];
```
Comment (edwardbrowncross) [inline source/server.js]: As in the client, this could be simpler with `.find`

File: source/server.js (hunk 11 above)
Comment (edwardbrowncross) [reply source/server.js]: But having said that, this function would be easier to test and easier to understand if it returned the scores rather than setting a global variable. Everywhere the global variable is used, this function is called right before anyway, so there would be little impact on the rest of the code

File: source/server.js [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+function isPositionTaken(_board, xPos, yPos) {
+  if (xPos === undefined || yPos === undefined) {
+    throw new Error('Argument undefined');
+  }
+
+  return _board[yPos][xPos] !== 0;
+}
+
+function dropToBottom(_board, xPos, freePos) {
```
Comment (edwardbrowncross) [inline source/server.js:56]: This function accepts an argument of `freePos`, but that argument isn't used except to hold a value on line 61 (and even then this is not necessary as you could `return j`). I suggest it should be removed
```suggestion
function dropToBottom(_board, xPos) {
```

File: source/server.js [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+app.post('/game/column/:column', (req, res) => {
+  const freePos = 0;
+  const row = dropToBottom(board, req.params.column, freePos);
+
+  if (row === -1) {
+    return board;
+  }
+
+  if (turnCount % 2 === 0) {
```
Comment (edwardbrowncross) [inline source/server.js:137]: There are a number of places where you are basically converting turn count into a number 1 or 2. If you wrote a function that took turn count and returned the player number of the current turn, there are several bits of code that would become tidier.

File: source/server.js [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+  res.json({
+    board,
+    player: turnCount % 2 === 0 ? 1 : 2,
+  });
+});
+
+app.post('/game/reset-game', (req, res) => {
+  res.json({
+    board: resetGame(board),
+    player: turnCount % 2 === 0 ? 1 : 2,
```
Comment (edwardbrowncross) [inline source/server.js]: `resetGame` sets the turn count to 0, so player should always be 1 here I believe.
```suggestion
    player: 1,
```

File: source/server.js [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+    });
+  } else if (verticalWinner !== -1) {
+    await updateScore(verticalWinner);
+    res.json({
+      winner: verticalWinner,
+      gameEnded: true,
+      scores,
+    });
+  } else {
+    res.json(-1);
```
Comment (edwardbrowncross) [inline source/server.js]: This endpoint sometimes sends back an object and sometimes a number. That makes it slightly harder to consume.

File: tests/connect4.test.js [hunk 17]
```diff
@@ -1,20 +1,19 @@
...
-    document.getElementById("row-0-column-0").style.backgroundColor = "blue";
-    each([
-        [document.getElementById("row-0-column-0"), "red"],
+const { drawBoard, setNextTurn, gameEnd } = require('../source/frontend/connect4.js');
 
-    ]).it("take turn with elements", (square, expected) => {
-        takeTurn(square);
-        expected(square.style.backgroundColor).toBe(expected);
-    })
+describe.skip('isTurnTaken', () => {
```
Comment (edwardbrowncross) [inline tests/connect4.test.js:3]: These tests are skipped. Why's that? Can the tests be removed or should they be enabled?

File: tests/server.test.js [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+  });
+
+  test('Returns false if an argument is undefined', () => {
+    expect(() => {
+      isPositionTaken(testBoard, undefined, 3);
+    }).toThrow('Argument undefined');
+  });
+});
+
+describe('findMatch', () => {
```
Comment (edwardbrowncross) [inline tests/server.test.js:71]: Looks like this is missing a happy path test (all four the same)

File: tests/server.test.js [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    const output = findMatch(1, undefined, undefined, undefined);
+    expect(output).toBe(false);
+  });
+});
+
+describe('droptoBottom', () => {
+  // Arange
+  const testBoard = [
+    [1, 1, 1, 1, 1, 1, 1],
+    [1, 1, 1, 1, 1, 0, 1],
```
Comment (edwardbrowncross) [inline tests/server.test.js:96]: The problem with this test data is it does not test whether it scans the column from the top or from the bottom. That is, it doesn't test whether it drops to bottom or drops to top. You would need two empty cells to test that.

File: tests/server.test.js [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+
+    // Act
+    const output = resetGame(testBoard);
+
+    // Assert
+    expect(output).toMatchObject(expectedOutput);
+  });
+});
+
+describe('horizontalWinCheck', () => {
```
Comment (edwardbrowncross) [inline tests/server.test.js:160]: It would be a good idea to test that it wont find a win if there isn't one. At the moment you could replace `horizontalWinCheck` with this and all these tests would still pass, even though it contains no logic.
```js
function horizontalWinCheck (board, player) {
    return player
}
```

File: source/server.js [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+    scores,
+  });
+});
+
+app.post('/game/column/:column', (req, res) => {
+  const freePos = 0;
+  const row = dropToBottom(board, req.params.column, freePos);
+
+  if (row === -1) {
+    return board;
```
Comment (edwardbrowncross) [inline source/server.js]: returning here doesn't send a response. It just exits the callback. You need to use `rest.json()` or `res.send()` or the client will experience a request that never comes back until it eventually times out.

File: source/frontend/functions.js [hunk 22]
```diff
@@ -0,0 +1,29 @@
+const turnCount = 0;
+let currentPlayer;
+
+const Game = {
```
Comment (edwardbrowncross) [inline source/frontend/functions.js]: Capital letters are normally reserved for classes.

**Review by edwardbrowncross — COMMENTED** [review]
This is looking really good generally. The code is quite easy to understand and is broken into clear functions of about the right size. There weren't many place where I was confused or surprised by your code, and that is a good measure of success.

You've also done a really good job of separating client and server using APIs.

You've made a good start on testing. In a real project, we'd want many more tests before merging this but the tests you have are pretty sound. It would be good for you to start thinking about edge cases and unusual inputs your functions could receive to test those too to make sure all behaviours are both tested and documented.

I've left some comments. Many of them are just small things around keeping the code tidy. Untidy code is a kind of tech debt that can make the code harder to maintain in the future. Taking a little time to remove unused code or unused inputs to functions, or refactoring code to remove duplication or complexity, pays off in the long run

One minor thing I couldn't put in the inline comments is that there are a lot of `.DS_Store` files committed, which come from MacOS. These shouldn't make it into the repo. You can create a `.gitignore` file to prevent this (look it up if you don't know what that means).
